### PR TITLE
MM-23395: Add websockets support for update group

### PR DESCRIPTION
**Pull request by catalintomai**
#### Summary
Ensuring that when we delete or update a MM group, the websocket event is triggered and we are updating the referenceable groups (et al) in Redux for group-mentions

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-23395 Note: this addresses items(events) 1-9 of the ticket - see ticket comments.

#### Related Pull Requests
redux: https://github.com/mattermost/mattermost-redux/pull/1113
webapp: https://github.com/mattermost/mattermost-webapp/pull/5245

Note: removed the mobile changes: mattermost/mattermost-mobile#4116 from this PR since they require extra consideration and are not blocking for the current set for webapp.

### CODE REVIEW

File: app/group.go [hunk 1]
```diff
@@ -32,11 +32,27 @@ func (a *App) CreateGroup(group *model.Group) (*model.Group, *model.AppError) {
...
+		messageWs.Add("group", newGroup.ToJson())
+		a.Publish(messageWs)
+	}
+
+	return newGroup, err
 }
 
 func (a *App) DeleteGroup(groupID string) (*model.Group, *model.AppError) {
-	return a.Srv().Store.Group().Delete(groupID)
+	newGroup, err := a.Srv().Store.Group().Delete(groupID)
```
Comment (mkraft) [inline app/group.go:47]:
```suggestion
	deletedGroup, err := a.Srv().Store.Group().Delete(groupID)
```

File: app/group.go [hunk 2]
```diff
@@ -32,11 +32,27 @@ func (a *App) CreateGroup(group *model.Group) (*model.Group, *model.AppError) {
...
+
+	if err == nil {
+		messageWs := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_RECEIVED_GROUP, "", "", "", nil)
+		messageWs.Add("group", newGroup.ToJson())
+		a.Publish(messageWs)
+	}
+
+	return newGroup, err
```
Comment (mkraft) [inline app/group.go:55]:
```suggestion
	return deletedGroup, err
```
...or `nil`.

File: app/group.go [hunk 3]
```diff
@@ -32,11 +32,27 @@ func (a *App) CreateGroup(group *model.Group) (*model.Group, *model.AppError) {
...
 }
 
 func (a *App) DeleteGroup(groupID string) (*model.Group, *model.AppError) {
-	return a.Srv().Store.Group().Delete(groupID)
+	newGroup, err := a.Srv().Store.Group().Delete(groupID)
+
+	if err == nil {
+		messageWs := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_RECEIVED_GROUP, "", "", "", nil)
```
Comment (mkraft) [inline app/group.go:50]: Should this be the same same websocket event as the used to update a group?

File: app/group.go (hunk 1 above)
Comment (catalintomai) [reply app/group.go]: updated.

File: app/group.go (hunk 3 above)
Comment (catalintomai) [reply app/group.go:50]: when we remove/delete a group we are actually update its delete_at timestamp, so once can look at it as the update of one of its properties. As we were discussing some time ago, the GetGroup* call(s) return the group if it exists in the table, "removed" or not, so the redux state is populated with both "removed" and not-removed groups. If the redux state would've only contain the non-removed ones, then I think that we would've need to add a specific DELETE event to actually remove the entry from the redux state when it is "removed" from the UserGroups table, with the contract that the redux state contains only non-removed groups.

**Comment by srkgupta** [comment]
QA to validate if all the points mentioned in the ticket https://mattermost.atlassian.net/browse/MM-23395 is working fine or not.

**Comment by srkgupta** [comment]
@catalintomai can you please resolve the merge conflicts so that this could be tested on top of the latest master.

**Comment by catalintomai** [comment]
> @catalintomai can you please resolve the merge conflicts so that this could be tested on top of the latest master.

@srkgupta , they are merged now. Thanks.

**Comment by srkgupta** [comment]
@catalintomai please ignore the issue #1 mentioned above. Actually I just tried this again, and editing a Group mention is working fine now and changes are reflected immediately via the Redux event. And also points #10 to #13 are also working fine. So the only thing which is not working is this:

A group is associated to a team
A group is disassociated to a team
A group is associated to a channel
A group is disassociated from a channel

These are also required to be handled, because, a team/channel can already be group synced. And then the sysadmin can decide to add one more group to that team/channel. In this case, the users in the team/channel don't see the list of group mentions correctly unless they manually refresh the browser/desktop.

**Comment by catalintomai** [comment]
> @catalintomai please ignore the issue #1 mentioned above. Actually I just tried this again, and editing a Group mention is working fine now and changes are reflected immediately via the Redux event. And also points #10 to #13 are also working fine. So the only thing which is not working is this:
> 
> A group is associated to a team
> A group is disassociated to a team
> A group is associated to a channel
> A group is disassociated from a channel
> 
> These are also required to be handled, because, a team/channel can already be group synced. And then the sysadmin can decide to add one more group to that team/channel. In this case, the users in the team/channel don't see the list of group mentions correctly unless they manually refresh the browser/desktop.

@srkgupta , you are correct, I failed to consider the case in which the channel/team is already group-synced and we are further linking groups to it. The latest update should contain the fix for the issue. ( have to run "npm run dev" in the redux folder to build the node_modules).

File: api4/channel.go [hunk 4]
```diff
@@ -322,7 +322,6 @@ func patchChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 	if c.Err != nil {
 		return
 	}
-
```
Comment (hahmadia) [inline api4/channel.go:325]: nit: we can probably add this back in

**Pull request merged**